### PR TITLE
Refactor promises

### PR DESCRIPTION
**Pull request by MartoYankov**
This is a proposed promises refactoring. I've done the git service only and will apply it to the other services once we agree if this is fine. My personal opinion is that bluebird is bad with typescript and it's better to use es6 despite it being slower.

**The initial discussion was wrong**
Initially, we discussed whether to go with bluebird or es6 and decided for bluebird. After some research and exploring the code I think our discussion was wrong. Basically, we didn't get what ES6 promises were and that they are not that different from the bluebird idea.

**The deferred antipattern**
The problem that I identified was that we are using what is called the "deferred antipattern". It's explained in this SO - https://stackoverflow.com/questions/23803743/what-is-the-explicit-promise-construction-antipattern-and-how-do-i-avoid-it . Additionally, reading through this article about ES6 promises (http://2ality.com/2014/10/es6-promises-api.html) confirmed to me that es6 and bluebird are not that different.

Basically, we should use bluebird and es6 promises the same way. The only notable difference is that bluebird has Promise.method() for synchronous code and promisify() for promisifying while es6 has new Promise() for both use cases.

**The problem with Bluebird**
Bluebird has a huge problem with TypeScript. In order to use it you have to globally override the native Promise type with the bluebird promise type and even then, we you use static bluebird method, you have to call them like this "Bluebird.all()" instead of "Promise.all()". Moreover, there is no notion in your project that the Promise object is actually the bluebird one. The cleanest way to add bluebird promises to your project is to add the @types/bluebird-global npm package to your project that automatically does this hack. This is a hack btw.

**The problem with es6 promises**
They are slower than bluebird - http://bluebirdjs.com/docs/benchmarks.html . They also lack some methods like Promise.props() in bluebird. We need such a method in the git service and I had to do it manually in the Utils class using a Map. Other than these things, there is no notable difference.

### CODE REVIEW

**Review by borisKarastanev — DISMISSED** [review]

**Comment by MIYankov** [comment]
Some further thoughts:
1. The arrow functions are always anonymous. This can potentially lead to harder stack trace to read since a lot of the functions in the promises are not named. However, I think this is mitigated by the TS catching a lot of errors. I also tested and the stack trace looks something like this: at getNpmPackageVersion.then.then - which is fine. It means the second then() after the named function.

2. The git service is currently not tested and is not testable? The tests now fully stub it. If we want to test the git service in isolation, it will be hard to stub the request calls. I think the best approach would be to isolate the request calls in a different service (data?) that has only them and stub it.

File: lib/services/git-service.ts [hunk 1]
```diff
@@ -94,43 +87,42 @@ export class GitService implements IGitService {
...
                 }
             }
-            Promise.all(promises)
-                .then(() => {
-                    resolve(content);
-                })
-                .catch((error) => {
-                    reject(error);
+
+            resolve(promisesMap);
```
Comment (manoldonev) [inline lib/services/git-service.ts]: This does not seem right to me -- we are resolving something but then we are continuing execution with the then(...) statement?

Also we are not handling the rejection case?

File: lib/services/git-service.ts [hunk 2]
```diff
@@ -94,43 +87,42 @@ export class GitService implements IGitService {
...
+        })
+            .then((promisesMap: Map<string, Promise<any>>) => {
+                return util.promiseAllMap(promisesMap);
```
Comment (manoldonev) [inline lib/services/git-service.ts]: Do we need explicit Promise creation in this method at all?

I might be missing something but isn't this what we are looking for:
``` ts
    private getResourcesFromSource(templateName: string, assetDictionary: any, versionTag: string): Promise<any> {
        const promisesMap: Map<string, Promise<any>> = new Map();

        for (const key in assetDictionary) {
            if (assetDictionary.hasOwnProperty(key)) {
                promisesMap.set(key, this.getResource(templateName, versionTag, assetDictionary[key]));
            }
        }

        return util.promiseAllMap(promisesMap);
    }
```

File: lib/services/git-service.ts [hunk 3]
```diff
@@ -139,26 +131,30 @@ export class GitService implements IGitService {
...
-                    if (response.body && response.body["dist-tags"] && response.body["dist-tags"].latest) {
-                        version = "v" + response.body["dist-tags"].latest;
-                    }
-
-                    resolve(version);
-                })
-                .catch((error: any) => {
-                    // fallback to using the master (latest version)
-                    resolve("master");
+            resolve(templateName);
```
Comment (manoldonev) [inline lib/services/git-service.ts]: Why resolve here?

Comment (manoldonev) [reply lib/services/git-service.ts]: I think no explicit Promise construction is necessary here either:

``` ts
    private getNpmPackageVersion(templateName: string): Promise<any> {
        if (templateName.indexOf("tns-") !== 0) {
            templateName = "tns-" + templateName;
        }

        return util.request({
            method: "GET",
            uri: util.format("https://registry.npmjs.org/%s/", templateName),
            json: true,
            resolveWithFullResponse: true,
            headers: util.defaultHeaders
        })
            .then((response: any) => {
                let version = "master";
                if (response.body && response.body["dist-tags"] && response.body["dist-tags"].latest) {
                    version = "v" + response.body["dist-tags"].latest;
                }

                return version;
            })
            .catch((error: any) => {
                // fallback to using the master (latest version)
                return "master";
            });
    }
```

File: lib/services/git-service.ts [hunk 4]
```diff
@@ -43,34 +39,31 @@ export class GitService implements IGitService {
...
-                })
-                .catch((downloadPageError: any) => {
-                    reject(downloadPageError);
                 });
-        });
+            });
     }
 
-    private getPageTemplatesBaseUrl(flavor: string) {
+    private getPageTemplatesBaseUrl(flavor: string): Promise<any> {
```
Comment (manoldonev) [inline lib/services/git-service.ts]: This method is synchronous, there does not seem any point in making it async to me.

Comment (manoldonev) [reply lib/services/git-service.ts]: In `clonePageTemplate(...)` where it is used we are explicitly creating another Promise anyway so we do not need this async?

**Comment by MIYankov** [comment]
After some discussion with @manoldonev we agreed:
1. No to bluebird.
2. Refactor the deferred antipattern.
3. Use promises only for async calls.
4. Use the Promise constructor for promisifying callback based event handlers.
5. Move all code with parsing logic away from the git service, so that it becomes testable.
6. Refactor util - remove the modules and import them where needed, move the pageExists to the pages service.

File: lib/services/git-service.ts [hunk 5]
```diff
@@ -2,34 +2,34 @@ import { Config } from "../shared/config";
 import util from "../shared/util";
 
 export class GitService implements IGitService {
+    private defaultHeaders = {
```
Comment (manoldonev) [inline lib/services/git-service.ts]: I think this should either be static or declared outside the class.

**Review by manoldonev — APPROVED** [review]

**Review by lini — APPROVED** [review]

**Pull request merged**